Add support for terminal hyperlinks

hoelzro opened the pull request:
**Description**
Adds support for OSC 8 terminal links

**Related Issue**
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
Resolves #812

**How Has This Been Tested?**
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
I conducted manual tests for this

**Types of changes**
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation
- [ ] Chore (Related to CI or Packaging to platforms)

hoelzro committed:
feat: Add ground work for supporting hyperlinks to iostreams

See GH #812

See https://gist.github.com/egmontkob/eb114294efbcd5adb1944c9f3cb5feda
for technical details behind the hyperlink escape sequence

For now hyperlinks are enabled as long as long as standard output is a
TTY - we'll want to add some additional detection/flags down the road

hoelzro committed:
feat: use terminal hyperlinks when listing issues/MRs

See GH #812

hoelzro commented:
I submitted this early for quicker feedback on the code - I'm not a huge fan of the `if displayHyperlinks { return makeHyperlink(...) } else { return ... }`, but I wanted to get something pushed for others to look at.  Would it be overkill to have `MakeHyperlink` just return the display string if hyperlinks are _not_ wanted, so display code making use of hyperlinks doesn't need to dispatch on `DisplayHyperlinks` itself?

Oh, and I'm not attached to the method names I added - they should probably stand to be improved :smile:

hoelzro committed:
feat(hyperlinks): Add display_hyperlinks config option

hoelzro committed:
chore(hyperlinks): Clean up Hyperlink API a bit

hoelzro committed:
feat(hyperlinks): default hyperlinks to off

I think hyperlink display should default to off while the feature is new
so that users can play with it

hoelzro committed:
feat(hyperlinks): Implement FORCE_HYPERLINKS

Setting the FORCE_HYPERLINKS environment variable forces glab to render
hyperlinks when it can - otherwise, it falls back to display_hyperlinks
from the config, which tells glab whether or not to enable hyperlinks
based on whether or not we're outputting to a TTY.

hoelzro committed:
fix(hyperlink): accommodate hyperlink OSC in string stripping

hoelzro committed:
fix: Fix incorrect web_url in fixture data

hoelzro committed:
test(hyperlink): Add tests for hyperlinks in issues list

hoelzro committed:
test(hyperlink): Add tests for truncating text with hyperlink OSC

hoelzro committed:
fix(hyperlinks): accommodate hyperlinks in Truncate

hoelzro committed:
test(hyperlinks): Add tests for showing hyperlinks in mr list

hoelzro committed:
docs(hyperlink): Add docs for FORCE_HYPERLINKS and display_hyperlinks

profclems reviewed (commented):
Great work on this @hoelzro!

This is huge. I will take time this evening (my time 😄) to review

hoelzro committed:
fix(hyperlinks): simplify truncate algorithm

…while also fixing tableprinter tests, which expect that the first
grapheme provided to Truncate is always preserved

hoelzro commented:
@profclems Awesome, thanks! I just added docs and fixed the tableprinter tests, so it should be ready to roll if it looks good to you!

hoelzro commented:
@profclems Have you had a chance to look at this?  I know it's a pretty big PR - let me know if it would help to break up a bit! :sweat_smile:

profclems commented on pkg/iostreams/iostreams.go:
```diff
@@ -200,6 +203,32 @@ func (s *IOStreams) BackgroundColor() string {
+	case "always":
+		return true
+	case "auto":
+		return s.IsaTTY
```
The OSC doesn't seem to work well when piped to pager. This is what happens:
```sh
Showing 1 open issue in profclems/glab that match your search (Page 1)

8;;https://gitlab.com/profclems/glab/-/issues/278#2788;;  .gitconfig credential helper path does not work on Windows  (gh-cloned)  about 5 months ago
```


Maybe we should check if the result is piped to pager
```go
return s.IsaTTY && (s.pagerCommand == "" || s.pagerCommand == "cat")
```

hoelzro committed:
fix(hyperlinks): Deactivate hyperlinks in auto mode if we're writing to a pager

hoelzro commented:
@profclems Good catch! I just added a commit to fix that

profclems reviewed (approved):
This is great!
